cpp/include/cudf/table/experimental/row_operators.cuh

```diff
@@ -442,43 +446,92 @@ class device_row_comparator {
+                                             : cuda::std::pair(weak_ordering::GREATER, _depth);
+          }
+          // both nulls at arbitrary nesting
+          else if (is_left_list_null && is_right_list_null) {
+            if (l_def_level == r_def_level) { continue; }
+            // push less nested null to front
+            if (_null_precedence == null_order::BEFORE) {
+              return l_def_level < r_def_level ? cuda::std::pair(weak_ordering::LESS, _depth)
+                                               : cuda::std::pair(weak_ordering::GREATER, _depth);
+            }
+            // push less nested null to back
+            else {
+              return l_def_level < r_def_level ? cuda::std::pair(weak_ordering::GREATER, _depth)
+                                               : cuda::std::pair(weak_ordering::LESS, _depth);
+            }
+          }
+          // left list empty
+          else if (is_left_list_empty) {
+            // if right is null
+            if (is_right_list_null) {
+              // push lower nested null to front
+              if (_null_precedence == null_order::BEFORE) {
+                return r_def_level < l_def_level ? cuda::std::pair(weak_ordering::GREATER, _depth)
+                                                 : cuda::std::pair(weak_ordering::LESS, _depth);
+              }
+            }
+            // left empty list always before right leaf or null when null_order::AFTER
+            return cuda::std::pair(weak_ordering::LESS, _depth);
+          }
+          // right list empty
+          else if (is_right_list_empty) {
+            // if left list is null
+            if (is_left_list_null) {
+              // let lower nested null be in front
+              if (_null_precedence == null_order::BEFORE) {
+                return l_def_level < r_def_level ? cuda::std::pair(weak_ordering::LESS, _depth)
+                                                 : cuda::std::pair(weak_ordering::GREATER, _depth);
+              }
+            }
+            // right empty leaf always before left leaf or null when null_order::AFTER
+            return cuda::std::pair(weak_ordering::GREATER, _depth);
+          }
+          // col.is_null(element_index) is not reliable after the list has been unraveled
+          // compare left null to right leaf
+          else if (is_left_list_null && r_def_level == r_max_def_level) {
+            return _null_precedence == null_order::BEFORE
+                     ? cuda::std::pair(weak_ordering::LESS, _depth)
+                     : cuda::std::pair(weak_ordering::GREATER, _depth);
+          }
+          // compare right null to left leaf
+          else if (is_right_list_null && l_def_level == l_max_def_level) {
+            return _null_precedence == null_order::BEFORE
+                     ? cuda::std::pair(weak_ordering::GREATER, _depth)
+                     : cuda::std::pair(weak_ordering::LESS, _depth);
+          }
```

Comment:
I wonder if all these if-else can be simplified somehow? They are too long, too many branches which is very bad in GPU processing.

Reply:
Yes, we might just scrap this complicated logic for now and go with what Spark expects in terms of ordering. That will be much simpler looking.